### DISCUSSION TITLE
Teams feature cannot manage authorization above the team's scope

### DISCUSSION THREAD

**Discussion by mohamedds-12** [spatie/laravel-permission#2349] [Q&A] [answered]
In my application, an organization has multiple events, and a user can have multiple roles in different events, Until now it's all good I can just work with teams considering the event is the team, But the problem is that I have some roles that are focused to the organization ex: `Owner, Admin` and they're not related to the event. And therefore the previous approach cannot fix my problem cuz it's limited to users having roles in an event (team).

**Is there any way I can assign different roles in different events (teams) for the user, along with the ability to assign some roles outside of the events for some users ?**

**Comment by parallels999** [comment]
Yes, read documentation
[teams-permissions#content-roles-creating](https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions#content-roles-creating)

**Reply by mohamedds-12** [reply]
That suggests that each time I create a team, I should get the super admin user and assign it the **super-admin** role in that team, so basically, he should be present in all teams. That's useful if you have one super admin user that you have a way to get it, but in my case I have a lot of them with different global roles (ex: `Owner, Admin, ...`) I can't even tell who they are (cuz they can get created).
On top of that in my case, the organization gets created first, so at first, when I add an `Owner` for example to which team should I assign him ( considering that no event was created yet)?

**Reply by parallels999** [reply]
It looks as if this package does not meet your requirements, have you checked out any of the alternatives?
>I can't even tell who they are (cuz they can get created)

When they are created, somebody most assign it, also i don't point you `super-admin` part, i point you [roles-creating](https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions#content-roles-creating)

**Reply by mohamedds-12** [reply]
>i point you [roles-creating](https://spatie.be/docs/laravel-permission/v5/basic-usage/teams-permissions#content-roles-creating)

Role creating is not the problem. The problem is that I have some resources outside of the team. So how can you check for the permissions that they can/cannot do on some resources outside of the team when there is no team_id yet, cuz you're outside of the team?

**Reply by erikn69** [reply]
> it looks as if this package does not meet your requirements, have you checked out any of the alternatives?

Maybe, [bouncer](https://github.com/JosephSilber/bouncer)

**Reply by mohamedds-12** [reply]
> > it looks as if this package does not meet your requirements, have you checked out any of the alternatives?
> 
> Maybe, [bouncer](https://github.com/JosephSilber/bouncer)

I did take a look at it before but not a deep one, I will take a deeper look at it and see.

**Reply by yungifez** [reply]
Hello, did you solve this problem?

**Reply by mohamedds-12** [reply]
Hi, sorry I forgot that I've posted a thread here too !, Yes I've solved the problem and I will post it right away.

**Comment by mohamedds-12** [comment, marked as answer]
Eventually, after searching and asking other senior engineers I ended up going with the **polymorphic solution** : 

>  I've tried going for a polymorphic relationship which basically means a user can have a role either in an event or an organization like so: user_id, role_id, entity_id, entity_type, But it's a lot of work and I don't know if it's the ideal solution.

It's a little complicated, but here is the breakdown of the steps :

1. **Polymorphic Relationship**:
   - Establish a polymorphic relationship between the `teams` table and the `events` and `organizations` tables. This allows a team to belong to either an event or an organization.

2. **Creating Teams for Organizations**:
   - Create a team for each organization. This team will serve as a container for users associated with that organization.

3. **Inviting Users and Assigning Roles**:
   - Invite users to join the organization's team.
   - Assign users roles such as Owner or Organization Admin at the organization level.

4. **Creating Teams for Events**:
   - When creating an event within an organization, create a new team specifically for that event.

5. **Assigning Teams to Events**:
   - Establish a link between the event and its corresponding team through the polymorphic relationship. This ensures that permissions for the event can be managed separately from the organization.

6. **Managing Event Permissions**:
   - Assign users from the organization's team roles like Event Admin or Moderator in the event's team. This ensures that users who have permissions at the organization level also have the necessary permissions for the events they are involved in.

7. **Inviting Users to Event Teams**:
   - Invite users to be part of the event's team.
   - Assign them roles like Event Admin, Moderator, etc., based on their responsibilities within the event.
